openstates/me/people.py

```diff
@@ -31,75 +31,99 @@ def scrape(self, chamber=None):
+            'seat_no': 19,
+            'house_leadership': 20,
+            'committees': 21
+        }
 
-            # Get the photo url.
-            html = self.get(leg_url).text
-            doc = lxml.html.fromstring(html)
-            doc.make_links_absolute(leg_url)
-            (photo_url,) = doc.xpath('//img[contains(@src, ".jpg")]/@src')
+        url = (
+            'http://legislature.maine.gov/house/128house.xlsx'
+        )
+        fn, result = self.urlretrieve(url)
 
-            # Add contact information from personal page
-            office_address = re.search(
-                r'<B>Address:  </B>(.+?)\n?</?P>', html, re.IGNORECASE).group(1)
+        wb = xlrd.open_workbook(fn)
+        sh = wb.sheet_by_index(0)
 
-            office_email = doc.xpath(
-                '//a[starts-with(@href, "mailto:")]/text()')
-            business_phone = re.search(
-                r'<B>Business Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
-            home_phone = re.search(
-                r'<B>Home Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
-            cell_phone = re.search(
-                r'<B>Cell Telephone:  </B>(.+?)</?P>', html, re.IGNORECASE)
+        for rownum in range(1, sh.nrows):
+            # get fields out of mapping
+            d = {}
+            for field, col_num in mapping.items():
+                try:
+                    d[field] = str(sh.cell(rownum, col_num).value).strip()
+                except UnicodeEncodeError:
+                    # str typecasting will raise error for non-ascii characters.
+                    d[field] = sh.cell(rownum, col_num).value.encode('utf-8').strip()
+                except IndexError:
+                    # This col_num doesn't exist in the sheet.
+                    pass
+
+            district_number = d['district']
+            district_name = d['legal_res']
+
+            first_name = d['first_name']
+            middle_name = d['mid_name']
+            last_name = d['last_name']
+            jr_sr = d['jr_sr']
+
+            member_name = " ".join((first_name, middle_name, last_name, jr_sr))
+            member_name = re.sub(r'\s+', ' ', member_name).strip()
+
+            party = d['party']
+            party = _party_map[party]
+
+            # Determine legislator's URL
+            leg_url = 'http://legislature.maine.gov/house/hsebios/'
+            leg_url += last_name[:4].lower() + first_name[0].lower()
+            if len(middle_name) > 0:
+                leg_url += middle_name[0].lower()
+            else:
+                leg_url += '_'
+            leg_url += '.htm'
+
+            # Determine legislator's URL to get their photo
+            photo_url = 'http://legislature.maine.gov/house/photo128/'
+            photo_url += last_name[:4].lower() + first_name[0].lower()
+            if len(middle_name) > 0:
+                photo_url += middle_name[0].lower()
+            else:
+                photo_url += '_'
+            photo_url += '.jpg'
```

Comment:
I'd suggest the same pattern as above, to use `self.head(photo_url)` to confirm that this photo actually exists.

Reply:
Also, it seems like you could cut down on code, if the same `last_name[:4].lower() + first_name[0].lower() + (middle_name[0].lower() or '_')` exists for both leg _and_ photo URLs.